sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ColumnResolutionHelper.scala

```diff
@@ -369,4 +376,59 @@ trait ColumnResolutionHelper extends Logging {
       throws = true,
       allowOuter = allowOuter)
   }
+
+  private def resolveExpressionByPlanId(
+      e: Expression,
+      q: LogicalPlan,
+      allowOuter: Boolean): Expression = {
```

Comment:
if an attribute has a plan id, I think it should never be resolved to an outer reference

Reply:
ok. I think we can narrow the case to `UnresolvedAttribute` -> `AttributeReference`